### PR TITLE
Redefine function definitions for TSVECTOR columns

### LINKED ISSUE #3798: Fix TSVECTOR trigger definitions for search
Searching for common names with apostrophes such as "o'hare" returns results for what seems to be all contributions containing "o" instead of the expected "o'hare".

image

Is there a correct way of escaping passed quotes when using the API (where the same issues seems to apply)?

### PR DESCRIPTION
**Pull request by jason-upchurch**
## Summary (required)

- Resolves #3798 

redefine `fec_fitem_sched_b_insert()` such that `to_tsvector` splits on (and removes) non-word characters when generating `recipient_name_text` column.

## How to test the changes locally

- [ ] check out this branch
- [ ] run pytest
- [ ] Test migrations with drop/create cfdm_test and `invoke create_sample_db`
- [ ] query `cfdm_test` for tables/columns identified in https://github.com/fecgov/openFEC/issues/3798#issuecomment-501311701 and verify no special characters in the `<col_nam_prefix>_text` columns.
- [ ] for any `varchar` column that has a corresponding `tsvector` column, searching the `varchar` column in the form `select * from <table> where <column> @@ to_tsquery('part_1:* & part_2:* & ... & part_n: *');` should return that `varchar` column data. 
- [ ] Additionally, if the `varchar` column data is parsed to split on and remove any nonword characters, i.e., contain only [a-zA-Z0-9] where spaces and special characters split the string into words, that parsing should match the contents of the `tsvector` column (locally `tsvector` entries may have stems of full words rather than the full word itself.

## Impacted areas of the application
List general components of the application that this PR will affect:

-  all places where 'disclosure.fec_fitem_sched_b.recipient_name_text` is in use

### CODE REVIEW

File: webservices/utils.py [hunk 1]
```diff
@@ -240,16 +240,10 @@ def extend(*dicts):
...
-    return search string for tsquery containing the portion of each of the entered words
-    up to the first non-word character
-    """
-    # get list of entered words (in a single field)
-    words = re.split(r'\s+', text)
-    # get the portions of the words up to the first non-word character
-    word_parts = [re.match(r'(.*?)\W', word).group()[:-1] if re.match(r'(.*?)\W', word) else word for word in words]
-    # remove any empty strings from word_parts list and return the words in a formatted string
-    return ' & '.join([word_part + ':*' for word_part in word_parts if word_part])
+    return ' & '.join([
```
Comment (lbeaufort) [inline webservices/utils.py]: Let's add a docstring to explain the new/old functionality

Comment (jason-upchurch) [reply webservices/utils.py]: thanks @lbeaufort! Done!!

**Comment by fecjjeng** [comment]
@jason-upchurch 
What is the purpose of the following code?  The column election_cycle in fec_fitem_sched_b had been renamed to two_year_transaction_period long time ago.  


DO $$
BEGIN
    EXECUTE format('alter table disclosure.fec_fitem_sched_b rename column election_cycle to two_year_transaction_period');
    EXCEPTION 
             WHEN undefined_column THEN 
                null;
             WHEN others THEN 
                RAISE NOTICE 'some other error: %, %',  sqlstate, sqlerrm;  
END$$;

**Comment by fecjjeng** [comment]
Also, it is not necessary to drop the trigger and recreate it to pick up the updated definition of the trigger function.

**Comment by jason-upchurch** [comment]
> @jason-upchurch
> What is the purpose of the following code? The column election_cycle in fec_fitem_sched_b had been renamed to two_year_transaction_period long time ago.
> 
> DO $$
> BEGIN
> EXECUTE format('alter table disclosure.fec_fitem_sched_b rename column election_cycle to two_year_transaction_period');
> EXCEPTION
> WHEN undefined_column THEN
> null;
> WHEN others THEN
> RAISE NOTICE 'some other error: %, %', sqlstate, sqlerrm;
> END$$;
> 
> Same situation with
> adding column creditor_debtor_name_text to fec_fitem_sched_d
> adding column payee_name_text to fec_fitem_sched_f

@fecjjeng my approach here was to find the most recent migration file related to each table and copy/paste verbatim that code, while only updating the `to_tsvector` arguments. So that code previously existed in an earlier migration file. Is it ok to delete it from this migration file? (as well as adding column creditor_debtor_name_text to fec_fitem_sched_d
adding column payee_name_text to fec_fitem_sched_f?)

**Comment by jason-upchurch** [comment]
> Also, it is not necessary to drop the trigger and recreate it to pick up the updated definition of the trigger function.

@fecjjeng would it be ok/your preference if I go through the new migration file and remove all instances of the `drop trigger`?

**Comment by fecjjeng** [comment]
@jason-upchurch I can explain to you a bit more later.  I also have question on the real change you made.

**Comment by fecjjeng** [comment]
Run the following steps to test this PR:
check out the branch, 
- drop/create cfdm_test and invoke create_sample_db.  Migration file V0142 finished without error. 
- Visually checked the migration file to confirm the un-wanted code described above had been removed.
- run pytest, finished successfully.
- uncomment app.config['SQLALCHEMY_ECHO'] = True in rest.py
Verified the query output regarding the to_tsquery is as followed:
disclosure.fec_fitem_sched_b.recipient_name_text @@ to_tsquery(%(recipient_name_text_1)s)
'recipient_name_text_1': 'amazon:* & com:*'
- -- amazon.com
-- 34 rows ("AMAZON COM INC")
select recipient_nm, recipient_name_text
FROM disclosure.fec_fitem_sched_b_2019_2020
WHERE recipient_name_text @@ to_tsquery('amazon:* & com:*') 

-- 525 rows (in addition to "AMAZON COM INC", got "AMAZON.COM, INC.", "AMAZON.COM LLC", etc.)
select recipient_nm, recipient_name_text, to_tsvector(regexp_replace(recipient_nm, '[^a-zA-Z0-9]', ' ', 'g'))
FROM disclosure.fec_fitem_sched_b_2019_2020
WHERE to_tsvector(regexp_replace(recipient_nm, '[^a-zA-Z0-9]', ' ', 'g')) @@ to_tsquery('amazon:* & com:*') 
-- return t
select to_tsvector(regexp_replace('I like to shop at amazon.com', '[^a-zA-Z0-9]', ' ', 'g')) @@ to_tsquery('amazon:* & com:*') 

-- return f
select to_tsvector(regexp_replace('I like to shop at amazon cloud', '[^a-zA-Z0-9]', ' ', 'g')) @@ to_tsquery('amazon:* & com:*') 

With the above test results, approved this PR.

**Pull request merged**